src/CodeGen_LLVM.cpp

```diff
@@ -4261,17 +4270,94 @@ void CodeGen_LLVM::visit(const Evaluate *op) {
 }
 
 void CodeGen_LLVM::visit(const Shuffle *op) {
+    vector<Value *> vecs;
+    for (const Expr &e : op->vectors) {
+        vecs.push_back(codegen(e));
+    }
+
     if (op->is_interleave()) {
-        vector<Value *> vecs;
-        for (Expr i : op->vectors) {
-            vecs.push_back(codegen(i));
-        }
         value = interleave_vectors(vecs);
     } else {
-        vector<Value *> vecs;
-        for (Expr i : op->vectors) {
-            vecs.push_back(codegen(i));
+        // If the even-numbered indices equal the odd-numbered
+        // indices, only generate one and then do a self-interleave.
+        for (int f : {4, 3, 2}) {
+            bool self_interleave = op->indices.size() % f == 0;
+            for (size_t i = 0; i < op->indices.size(); i++) {
+                self_interleave &= (op->indices[i] == op->indices[i - (i % f)]);
+            }
+            if (self_interleave) {
+                vector<int> sub_indices;
+                for (size_t i = 0; i < op->indices.size(); i += f) {
+                    sub_indices.push_back(op->indices[i]);
+                }
+                Expr equiv = Shuffle::make(op->vectors, sub_indices);
+                value = codegen(equiv);
+                if (f == 2) {
+                    value = interleave_vectors({value, value});
+                } else if (f == 3) {
+                    value = interleave_vectors({value, value, value});
+                } else {
+                    value = interleave_vectors({value, value});
+                    value = interleave_vectors({value, value});
```

Comment:
I think we should let `interleave_vectors` handle this. Maybe there is some backend where there is a better implementation of `interleave_vectors({value, value, value, value})`? And if not, the implementation should just do this anyways.

And on that note, maybe just write this whole if as `interleave_vectors(std::vector<Value*>(f, value))`